File: modules/accelerators/contextualization/cdf_discovery/ui/src/utils/treeNodeIds.ts
/** Canonical Discovery tree node ids (mirror ``ui/server/tree_node_ids.py``). */

export const CONNECTION_ROOT = "connection";
export const DATA_ROOT = "data";
export const DATA_SAVED_QUERIES = "data:sq";
export const TRANSFORM_ROOT = "transform";
export const TRANSFORM_PIPELINES = "transform:pipelines";
export const TRANSFORM_PIPELINE_PREFIX = "transform:pipeline:";
export const TRANSFORM_TEMPLATES = "transform:templates";
export const TRANSFORM_TEMPLATE_PREFIX = "transform:template:";
export const TRANSFORM_WORKFLOW_PREFIX = "transform:workflow:";
export const FUSION_ROOT = "fusion";
export const FUSION_DM_ROOT = "fusion:dm";
export const FUSION_SPACES = "fusion:spaces";
export const FUSION_ADMIN = "fusion:admin";
export const FUSION_GROUPS = "fusion:admin:groups";
export const FUSION_INTEGRATION_ROOT = "fusion:integration";
export const GOVERNANCE_ROOT = "gov";
export const GOVERNANCE_SPACES = "gov:spaces";
export const GOVERNANCE_GROUPS = "gov:groups";
export const EXTRACT_ROOT = "extract";
export const MONITOR_ROOT = "monitor";

/** Sibling order under ``connection`` (tree root label shows the CDF project). */
export const CONNECTION_ROOT_CHILD_ORDER = [
  DATA_ROOT,
  FUSION_ROOT,
  GOVERNANCE_ROOT,
  EXTRACT_ROOT,
  TRANSFORM_ROOT,
  MONITOR_ROOT,
] as const;

export function dedupeNodeIds(nodeIds: readonly string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const nodeId of nodeIds) {
    const nid = nodeId.trim();
    if (!nid || seen.has(nid)) continue;
    seen.add(nid);
    out.push(nid);
  }
  return out;
}

File: modules/accelerators/contextualization/cdf_discovery/ui/server/tree_node_ids.py
"""Canonical Discovery tree node ids."""

from __future__ import annotations

CONNECTION_ROOT = "connection"
DATA_ROOT = "data"
DATA_SAVED_QUERIES = "data:sq"
TRANSFORM_ROOT = "transform"
TRANSFORM_PIPELINES = "transform:pipelines"
TRANSFORM_PIPELINE_PREFIX = "transform:pipeline:"
TRANSFORM_TEMPLATES = "transform:templates"
TRANSFORM_TEMPLATE_PREFIX = "transform:template:"
TRANSFORM_WORKFLOW_PREFIX = "transform:workflow:"
FUSION_ROOT = "fusion"
FUSION_DM_ROOT = "fusion:dm"
FUSION_SPACES = "fusion:spaces"
FUSION_ADMIN = "fusion:admin"
FUSION_GROUPS = "fusion:admin:groups"
FUSION_INTEGRATION_ROOT = "fusion:integration"
GOVERNANCE_ROOT = "gov"
GOVERNANCE_SPACES = "gov:spaces"
GOVERNANCE_GROUPS = "gov:groups"
EXTRACT_ROOT = "extract"
MONITOR_ROOT = "monitor"

# Sibling order under ``connection`` (tree root label shows the CDF project).
CONNECTION_ROOT_CHILD_ORDER = (
    DATA_ROOT,
    FUSION_ROOT,
    GOVERNANCE_ROOT,
    EXTRACT_ROOT,
    TRANSFORM_ROOT,
    MONITOR_ROOT,
)
